Fixes to xfem 3d cutter mesh (#32780)
idaholab/moose: modules/xfem/doc/content/source/outputs/XFEMCutMeshOutput.md

## Diagnosing XFEM Cutting Errors with `execute_on = xfem_mark`

When an XFEM run aborts inside the EFA cut-marking / fragment update phase (for example with errors such as `EFAelement3D::getMasterInfo, cannot find the given EFANode`), the failure happens *before* the timestep's normal output is written, so the standard `final` or `timestep_end` Exodus snapshot does not capture the state of the cutter mesh that caused the error.

Setting [!param](/Outputs/XFEMCutMeshOutput/execute_on) to `xfem_mark` on the `XFEMCutMeshOutput` writes the cutter mesh at the XFEM marking phase, which runs *before* the fragment update.  The resulting Exodus file therefore reflects the cutter geometry that XFEM is about to apply.  Viewing this file in ParaView lets you see exactly where the crack-growth surface is at the moment of the abort -- often the geometry reveals overlapping cut surfaces, near-degenerate triangles, or an unexpected crack-front location that drove the failure.

> This output is purely diagnostic; remove or revert to `final` once the underlying input is fixed, since it produces an additional Exodus file per cut iteration.

Reviewer: The imperative has no object and the pronoun has no clear antecedent: "remove" what, and "it" reads as pointing at "the underlying input" rather than at the diagnostic setting. Name the setting in both places.

```suggestion
This setting is purely diagnostic.  Set [!param](/Outputs/XFEMCutMeshOutput/execute_on) back to `final` once the underlying input is fixed, because `xfem_mark` writes an additional Exodus file on every cut iteration.
```

Author: Done.

_Prepared with Codex, powered by GPT-5._

> This setting is purely diagnostic.  Set [!param](/Outputs/XFEMCutMeshOutput/execute_on) back to `final` once the underlying input is fixed, because `xfem_mark` writes an additional Exodus file on every cut iteration.

Reviewer: `final` is not the default `execute_on` for an output (the default is `initial timestep_end`), so "set it back to `final`" prescribes an output cadence the reader's input may never have had. Point them back to whatever the input used before the diagnostic change.

```suggestion
This setting is purely diagnostic.  Restore [!param](/Outputs/XFEMCutMeshOutput/execute_on) to the value the input used before once the underlying problem is fixed, because `xfem_mark` writes an additional Exodus file on every cut iteration.
```

Author: Done.

_Prepared with Claude Opus 4.8 via Claude Code._

## Overview

> This class outputs the lower-dimensional mesh used to define cuts with XFEM.  It outputs the mesh in the Exodus format and outputs it in the original (undeformed) configuration.  The `XFEMCutMeshOutput` creates a file using the naming convention [!param](/Outputs/XFEMCutMeshOutput/file_base)_XFEMCutMeshOutput.e.  This class outputs a new Exodus file at every timestep with an extension that matches the convention used by MOOSE for other modified meshes (i.e. *.e-s0001, *.e-s0002, etc.),  because it assumes that the cutting mesh is always adding new nodes and elements as the crack grows.  This class is currently only implemented to be operable with the MeshCut classes user objects ([CrackMeshCut3DUserObject.md], [MeshCut2DFractureUserObject.md] and [MeshCut2DFunctionUserObject.md]) but will eventually include all cutter user objects deriving off a GeometricCutUserObject.

Reviewer: "the MeshCut classes user objects" does not parse -- "classes" and "user objects" are stacked with nothing joining them. Drop "classes".

```suggestion
This class outputs the lower-dimensional mesh used to define cuts with XFEM.  It outputs the mesh in the Exodus format and outputs it in the original (undeformed) configuration.  The `XFEMCutMeshOutput` creates a file using the naming convention [!param](/Outputs/XFEMCutMeshOutput/file_base)_XFEMCutMeshOutput.e.  This class outputs a new Exodus file at every timestep with an extension that matches the convention used by MOOSE for other modified meshes (i.e. *.e-s0001, *.e-s0002, etc.),  because it assumes that the cutting mesh is always adding new nodes and elements as the crack grows.  This class is currently only implemented to be operable with the MeshCut user objects ([CrackMeshCut3DUserObject.md], [MeshCut2DFractureUserObject.md] and [MeshCut2DFunctionUserObject.md]) but will eventually include all cutter user objects deriving off a GeometricCutUserObject.
```

Author: Done.

_Prepared with Codex, powered by GPT-5._

## Diagnosing XFEM Cutting Errors with `execute_on = xfem_mark`

When an XFEM run aborts inside the EFA cut-marking / fragment update phase (for example with errors such as `EFAelement3D::getMasterInfo, cannot find the given EFANode`), the failure happens *before* the timestep's normal output is written, so the standard `final` or `timestep_end` Exodus snapshot does not capture the state of the cutter mesh that caused the error.

> Setting `execute_on = xfem_mark` on the `XFEMCutMeshOutput` writes the cutter mesh at the XFEM marking phase, which runs *before* the fragment update.  The resulting Exodus file therefore reflects the cutter geometry that XFEM is about to apply.  Viewing this file in ParaView lets you see exactly where the crack-growth surface is at the moment of the abort -- often the geometry reveals overlapping cut surfaces, near-degenerate triangles, or an unexpected crack-front location that drove the failure.

Reviewer: The prose names the `execute_on` parameter but formats it as plain code. This page already uses `[!param]` for `file_base` on line 7; use the same form here so the reference links to the parameter table.

```suggestion
Setting [!param](/Outputs/XFEMCutMeshOutput/execute_on) to `xfem_mark` on the `XFEMCutMeshOutput` writes the cutter mesh at the XFEM marking phase, which runs *before* the fragment update.  The resulting Exodus file therefore reflects the cutter geometry that XFEM is about to apply.  Viewing this file in ParaView lets you see exactly where the crack-growth surface is at the moment of the abort -- often the geometry reveals overlapping cut surfaces, near-degenerate triangles, or an unexpected crack-front location that drove the failure.
```

Author: Done.

_Prepared with Codex, powered by GPT-5._

This output is purely diagnostic; remove or revert to `final` once the underlying input is fixed, since it produces an additional Exodus file per cut iteration.

## Example Input Syntax

> !listing test/tests/solid_mechanics_basic/face_crack_3d_xfem_scc.i block=Outputs replace=['execute_on=final','execute_on = xfem_mark']

Reviewer: This `replace=` pair will never fire, so the rendered listing still shows `final` while the section above tells the reader to use `xfem_mark`. With `block=` set, `InputListingCommand.getContent` runs the input through pyhit (`extractInputBlocks` -> `node.render()`), and hit's `Field::render` emits every parameter as `path() + " = "` (framework/contrib/hit/src/hit/parse.cc:669). The listing text is therefore `execute_on = final` with spaces, not the `execute_on=final` written in face_crack_3d_xfem_scc.i:52, and `str.replace` finds no match. Match the normalized form.

```suggestion
!listing test/tests/solid_mechanics_basic/face_crack_3d_xfem_scc.i block=Outputs replace=['execute_on = final','execute_on = xfem_mark']
```

Author: Done.

_Prepared with Codex, powered by GPT-5._

This class outputs the lower-dimensional mesh used to define cuts with XFEM.  It outputs the mesh in the Exodus format and outputs it in the original (undeformed) configuration.  The `XFEMCutMeshOutput` creates a file using the naming convention [!param](/Outputs/XFEMCutMeshOutput/file_base)_XFEMCutMeshOutput.e.  This class outputs a new Exodus file at every timestep with an extension that matches the convention used by MOOSE for other modified meshes (i.e. *.e-s0001, *.e-s0002, etc.),  because it assumes that the cutting mesh is always adding new nodes and elements as the crack grows.  This class is currently only implemented to be operable with the [MeshCut2DFractureUserObject.md] and [MeshCut2DFunctionUserObject.md] user objects but will eventually include all cutter user objects deriving off a GeometricCutUserObject.

## Diagnosing XFEM Cutting Errors with `execute_on = xfem_mark`

> When an XFEM run aborts inside the cut-marking / fragment update phase (for example with errors such as `one embedded node must be owned by 2 faces` or `In EFAelement3D::getMasterInfo, cannot find the given EFANode`), the failure happens *before* the timestep's normal output is written, so the standard `final` or `timestep_end` Exodus snapshot does not capture the state of the cutter mesh that caused the error.

Reviewer: The quoted error text `one embedded node must be owned by 2 faces` does not match any error string produced by the xfem module. The relevant error path, `EFAFragment3D::removeInvalidEmbeddedNodes` in the over-shared case, actually raises "... embedded node <id> is shared by <N> fragment faces in element <M> (expected 2). Non-manifold fragment topology, needs investigation." Anyone grepping logs for the quoted phrase from this doc will not find a match. The doc should quote the real message text.

Author: Done.

_Prepared with Codex, powered by GPT-5._

Setting `execute_on = xfem_mark` on the `XFEMCutMeshOutput` writes the cutter mesh at the XFEM marking phase, which runs *before* the fragment update.  The resulting Exodus file therefore reflects the cutter geometry that XFEM is about to apply.  Viewing this file in ParaView lets you see exactly where the crack-growth surface is at the moment of the abort -- often the geometry reveals overlapping cut surfaces, near-degenerate triangles, or an unexpected crack-front location that drove the failure.

This output is purely diagnostic; remove or revert to `final` once the underlying input is fixed, since it produces an additional Exodus file per cut iteration.

Example:

```
[Outputs]
  [xfemcutter]
    type = XFEMCutMeshOutput
    xfem_cutter_uo = cut_mesh
    execute_on = xfem_mark
  []

> []

Author: this should point to an input file

Author: Done.

_Prepared with Codex, powered by GPT-5._
